src/main/java/pl/edu/agh/firecell/storage/FileSystemStorage.java

```diff
@@ -0,0 +1,81 @@
+package pl.edu.agh.firecell.storage;
+
+import io.reactivex.rxjava3.disposables.Disposable;
+import io.reactivex.rxjava3.schedulers.Schedulers;
+import io.reactivex.rxjava3.subjects.PublishSubject;
+import io.reactivex.rxjava3.subjects.Subject;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import pl.edu.agh.firecell.model.State;
+import pl.edu.agh.firecell.storage.serialization.StateSerializer;
+
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Optional;
+
+public class FileSystemStorage implements Storage {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private final Subject<StateHolder> stateSubject = PublishSubject.create();
+    private final StateSerializer serializer;
+    private final Path path;
+
+    private Disposable subscription;
+
+    public FileSystemStorage(StateSerializer serializer, Path path) {
+        this.serializer = serializer;
+        this.path = path;
+    }
+
+
+    @Override
+    public void putState(State state, int index) {
+        stateSubject.onNext(new StateHolder(state, index));
+    }
+
+    @Override
+    public Optional<State> getState(int index) {
+        File file = getFile(index);
+        try (InputStream stream = new FileInputStream(file)) {
+            State state = serializer.parseFrom(stream);
+            return Optional.of(state);
+        } catch (FileNotFoundException e) {
+            return Optional.empty();
+        } catch (IOException e) {
+            e.printStackTrace();
+            return Optional.empty();
+        }
+    }
+
+    public void initialize() throws IOException {
```

Comment:
Why not call this function inside constructor?
We should not create objects not fully initialized, and  `FileSystemStorage` would actually not work properly when we forget to call this method from outside

Reply:
added initialization to constructor